feat: Fix for issue#663

jaishirole opened the pull request:
Signed-off-by: jaishirole <jaishirole@gmail.com>

## Description of changes
Fixes https://github.com/loopbackio/loopback-connector-mongodb/issues/663
Makes sure that fix does cover https://github.com/loopbackio/loopback-connector-mongodb/issues/610 too

1. Since Mongo Connection URLs can have commas in it, the URL class based on WHATWG standard in current state, is unable to parse these URLs. Though there exists a method to construct a MongoDB URL from the settings, the code change was needed to parse a MongoURL and explicitly insert database if one exists in the settings. This will happen only if
the url doesn't have a database specified and settings has it.

2. Please note that I have modified expected results of 3 existing unit test cases as the current Mongo version now returns a different error message than in the past.
3. The package-lock.json is added based on modifications seen on my system.
4. This is my first commit in Loopback project, looking forward to the feedback if something isn't done right way. Thanks!

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
<img width="559" alt="image" src="https://user-images.githubusercontent.com/37609734/164014315-e0e77067-ae44-4458-9ef9-80e57fbc935a.png">

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](https://loopback.io/doc/en/contrib/style-guide-es6.html)
- [x] Commit messages are following our [guidelines](https://loopback.io/doc/en/contrib/git-commit-messages.html)

achrinza commented on test/mongodb.test.js:
```diff
@@ -3372,6 +3372,143 @@ describe('mongodb connector', function() {
     });
   });
 
+  describe('Test processMongoDBUrl function', function() {
```
Feel free to ignore this:

A trick that we use in new tests in LB4  is to programmatically generate `it()` using a for-loop and a list(s) of valid value so that we can test every reasonable permutation. This can be considered out of scope for this PR as I think these tests should be sufficient for now.

jaishirole replied:
Good to know, thanks!

achrinza commented:
I suspect it's because of these lines:

https://github.com/loopbackio/loopback-connector-mongodb/blob/5936ba0cd74a2b7dd84bb0925e0a9636de4f5d85/setup.sh#L44

https://github.com/loopbackio/loopback-connector-mongodb/blob/5936ba0cd74a2b7dd84bb0925e0a9636de4f5d85/setup.sh#L49

The CI pipeline pins to v4.4:

https://github.com/loopbackio/loopback-connector-mongodb/blob/5936ba0cd74a2b7dd84bb0925e0a9636de4f5d85/.github/workflows/continuous-integration.yaml#L25

Compare to `latest` which AFAIK points to `5.0.7`.

Try changing the line in your local copy of `setup.sh` to point to `mongodb:4.4` instead.

achrinza commented on test/mongodb.test.js:
```diff
@@ -3372,6 +3372,143 @@ describe('mongodb connector', function() {
+      it('when no seetings db, but url has db, multiple hosts, user credentials and options', function() {
+        const url = 'mongodb://userid:password@db1.example.com:27017,db2.example.com:32667/testdb?authSource=admin&replicaSet=replset&readPreference=primary&ssl=true';
+        const database = '';
+        module.processMongoDBURL('', url).should.be.eql('mongodb://userid:password@db1.example.com:27017,db2.example.com:32667/testdb?authSource=admin&replicaSet=replset&readPreference=primary&ssl=true');
```
Could we add one test where the URL contains a trailing forward slash without a specified database (e.g. `mongodb://userid:password@db1.example.com/,mongodb://userid:password@db1.example.com/testdb`

jaishirole replied:
I already have this test case (trailing forward slash without specified database) already:
```
it("when settings has db, url doesn't have db, has slash, multiple hosts, user creds and no options", function() {
        const url = 'mongodb://userid:password@db1.example.com:27017,db2.example.com:32667/';
        const database = 'mydb';
        module.processMongoDBURL(database, url).should.be.eql('mongodb://userid:password@db1.example.com:27017,db2.example.com:32667/mydb');
      });
```
and with a database
```
it('when settings has db, url too has db, multiple hosts, user credentials and no options', function() {
        const url = 'mongodb://userid:password@db1.example.com:27017,db2.example.com:32667/testdb';
        const database = 'mydb';
        module.processMongoDBURL(database, url).should.be.eql('mongodb://userid:password@db1.example.com:27017,db2.example.com:32667/testdb');
      });
```

achrinza replied:
Ah, thanks! Didn't notice that test.

achrinza reviewed (commented):
Thanks for your PR, @jaishirole, most of it looks great! Just have a few comments :point_down:

jaishirole commented:
@achrinza Thanks much for the review comments. I'll incorporate the changes later today.

jaishirole commented on lib/mongodb.js:
```diff
@@ -332,13 +332,11 @@ MongoDB.prototype.connect = function(callback) {
-      self.settings.url = urlObj.toString();
+    // This is special processing if database is not part of url, but is in settings
+    if (self.settings.database && self.settings.url) {
+      if ((self.settings.url.indexOf('/' + self.settings.database) === -1) && (self.settings.url.indexOf('authSource=' + self.settings.database) === -1)) {
```
@achrinza My thinking is, if someone is including authSource, they do know of their database. If I change this to an 'OR', I see that for every mongoURL, the new logic will be invoked. I was trying to see how to narrow down calls into it and hence '&&' made sense as this lets us process the URL if url doesn't have a database specified (after `/`) and not even in `authSource` parameter.

jaishirole replied:
@achrinza This indeed is a flaky condition, let me think over and push a fix. The unit tests directly call the new function, hence I missed testing this part. Will take care of it in new commit. Can you trigger the workflow, to see other issues are taken care now?

jaishirole replied:
@achrinza Pushing a new commit with modified condition as well as adding a few more functional test cases that will execute the new function more.

jaishirole replied:
Locally passing test coverage post latest commit:
<img width="429" alt="image" src="https://user-images.githubusercontent.com/37609734/165822316-8328b208-7136-4f8c-90bf-5a234d4b4542.png">

achrinza commented on lib/mongodb.js:
```diff
@@ -2118,6 +2116,99 @@ MongoDB.prototype.rollback = function(tx, cb) {
+  if (mongoUrl) {
+    // 1. Know the protocol
+    let baseUrl = '';
+    if (mongoUrl.startsWith('mongodb:'))
+      baseUrl = 'mongodb://';
+    else if (mongoUrl.startsWith('mongodb+srv:'))
+      baseUrl = 'mongodb+srv://';
```
The changes LGTM. Thanks for also adding the additional logic to return the URL string as-is if it's not parseable.

achrinza commented:
Not sure why the coverage plummeted to zero, but the test runs themselves look ok.

jaishirole commented:
@achrinza Sorry, didn't get time during the day. Will look into tomorrow. I have an idea of what is missing.

jaishirole commented:
@achrinza Can you help trigger the Workflow to see what happens this time ?

achrinza commented:
Workflow triggered - Just a heads up, our workflow doesn't allow merging PRs with Merge Commits, so they'll need to be dropped and, if needed, replaced with a rebase instead.

jaishirole commented:
> Workflow triggered - Just a heads up, our workflow doesn't allow merging PRs with Merge Commits, so they'll need to be dropped and, if needed, replaced with a rebase instead.

Oh, didn't know. Looking into how to remove the merge commit from this PR now.. Any tips ?

jaishirole commented:
> Workflow triggered - Just a heads up, our workflow doesn't allow merging PRs with Merge Commits, so they'll need to be dropped and, if needed, replaced with a rebase instead.

Hi @achrinza , the merge commit is now dropped. Please review and let me know if I should fix anything more. Thanks!

achrinza commented:
Thanks for applying the changes and sorry for the delayed review; Everything LGTM :+1:

Just need to squash into a single Git Commit and it can be merged.

jaishirole commented:
> Thanks for applying the changes and sorry for the delayed review; Everything LGTM 👍
> 
> Just need to squash into a single Git Commit and it can be merged.

@achrinza I was on vacation and the squashing took unnecessarily longer. I'm sorry for the delay.
I've squashed all into one commit and when ran the UT locally, saw it passing:
<img width="494" alt="image" src="https://user-images.githubusercontent.com/37609734/171273151-e8568ff0-7e4f-49b0-9242-8cd6ebd91b74.png">

Can you please help with the next steps here? Thanks much for all your guidance so far!

gian1200 commented:
Hi all, is there any update on this? Thanks

marcioluis commented:
@dhmlau Hi all, is there any update on this? Thanks
I'm using https://www.npmjs.com/package/patch-package to fix this locally on my project as a workaround.

manarhusrieh commented:
I am worried that this package is no longer maintained. @dhmlau @achrinza is it going to be merged?

dhmlau commented:
@manarhusrieh, this PR was approved by @achrinza, so I'd assume it's good to go. 
But I see there are merging conflicts, @jaishirole, could you please resolve that? 

Question for @achrinza, this PR is for the `master` branch, but it has the `v6.x` label. Was it intended for 6.x too? If yes, we'd need another PR.

achrinza commented:
@dhmlau This should be merged into v7 now

manarhusrieh commented:
@achrinza can't this be merged to version 6? This is an issue with the package, not a new feature. We are currently using this PR package instead of the official one and it is working properly for more than a year.

dhmlau commented:
@manarhusrieh, could you please submit a new PR against the `6.x` branch? 
This PR is opened against the `master` branch which is 7.x. Thanks.

manarhusrieh commented:
@dhmlau OK, I'll prepare a PR against version 6. Thanks

manarhusrieh commented:
@dhmlau I have created the new PR for branch 6.x #741
